sentry/src/main/java/io/sentry/util/UrlUtils.java

```diff
@@ -0,0 +1,186 @@
+    final @Nullable String query =
+        extractQuery(url, queryParamSeparatorIndex, fragmentSeparatorIndex);
+    final @Nullable String fragment = extractFragment(url, fragmentSeparatorIndex);
+
+    return new UrlDetails(baseUrl, query, fragment);
+  }
+
+  private static @Nullable String extractBaseUrl(
+      final @NotNull String url,
+      final int queryParamSeparatorIndex,
+      final int fragmentSeparatorIndex) {
+    if (queryParamSeparatorIndex >= 0) {
+      return url.substring(0, queryParamSeparatorIndex).trim();
+    } else if (fragmentSeparatorIndex >= 0) {
+      return url.substring(0, fragmentSeparatorIndex).trim();
+    } else {
+      return url;
+    }
+  }
+
+  private static @Nullable String extractQuery(
+      final @NotNull String url,
+      final int queryParamSeparatorIndex,
+      final int fragmentSeparatorIndex) {
+    if (queryParamSeparatorIndex > 0) {
+      if (fragmentSeparatorIndex > 0 && fragmentSeparatorIndex > queryParamSeparatorIndex) {
+        return url.substring(queryParamSeparatorIndex + 1, fragmentSeparatorIndex).trim();
+      } else {
+        return url.substring(queryParamSeparatorIndex + 1).trim();
+      }
+    } else {
+      return null;
+    }
+  }
+
+  private static @Nullable String extractFragment(
+      final @NotNull String url, final int fragmentSeparatorIndex) {
+    if (fragmentSeparatorIndex > 0) {
+      return url.substring(fragmentSeparatorIndex + 1).trim();
+    } else {
+      return null;
+    }
+  }
+
+  private static @NotNull UrlDetails splitAbsoluteUrl(final @NotNull String url) {
+    try {
+      final @NotNull String filteredUrl = urlWithAuthRemoved(url);
+      final @NotNull URL urlObj = new URL(url);
+      final @NotNull String baseUrl = baseUrlOnly(filteredUrl);
+      if (baseUrl.contains("#")) {
+        // url considered malformed because it has fragment
+        return new UrlDetails(null, null, null);
+      } else {
+        final @Nullable String query = urlObj.getQuery();
+        final @Nullable String fragment = urlObj.getRef();
+        return new UrlDetails(baseUrl, query, fragment);
+      }
+    } catch (MalformedURLException e) {
+      return new UrlDetails(null, null, null);
+    }
+  }
+
+  private static @NotNull String urlWithAuthRemoved(final @NotNull String url) {
+    final @NotNull Matcher userInfoMatcher = AUTH_REGEX.matcher(url);
+    if (userInfoMatcher.matches() && userInfoMatcher.groupCount() == 3) {
+      final @NotNull String userInfoString = userInfoMatcher.group(2);
+      final @NotNull String replacementString =
+          userInfoString.contains(":") ? "[Filtered]:[Filtered]@" : "[Filtered]@";
+      return userInfoMatcher.group(1) + replacementString + userInfoMatcher.group(3);
+    } else {
+      return url;
+    }
+  }
+
+  private static @NotNull String baseUrlOnly(final @NotNull String url) {
+    final int queryParamSeparatorIndex = url.indexOf("?");
```

Comment:
`l`: Can't you use java built-in URL class here instead?
I haven't tried it but 
* [Url.getPath](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#getFile()) should reflect url with query/ref
* [Url.getRef](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#getRef()) should return the fragment
* [Url.getQuery](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#getQuery()) for the query params

Reply:
I guess you could reconstruct the base URL from parts retrieved from `URL` but imo that wouldn't be any better than the current code.

Reply:
Alright, then let's keep it this way. I just checked a few exotic URLs (like `example.com/?/path/?query=param`) and your solution provides better output than the URL api 👍